metrics/ci_integrations/lib/jenkins/adapter/jenkins_ci_client_adapter.dart

```diff
@@ -0,0 +1,182 @@
+  /// Throws [ArgumentError] if the given Jenkins client is `null`.
+  JenkinsCiClientAdapter(this.jenkinsClient) {
+    ArgumentError.checkNotNull(jenkinsClient, 'jenkinsClient');
+  }
+
+  @override
+  Future<List<BuildData>> fetchBuildsAfter(
+    String projectId,
+    BuildData build,
+  ) async {
+    ArgumentError.checkNotNull(build, 'build');
+
+    final buildingJob = await _fetchBuilds(
+      projectId,
+      limits: JenkinsQueryLimits.endBefore(0),
+    );
+    final lastBuild = buildingJob.lastBuild;
+    int numberOfBuilds = lastBuild.number - build.buildNumber;
+
+    if (numberOfBuilds <= 0) return [];
+
+    List<JenkinsBuild> builds;
+    while (builds == null) {
+      final buildingJob = await _fetchBuilds(
+        projectId,
+        limits: JenkinsQueryLimits.endAt(numberOfBuilds),
+      );
+
+      final _builds = buildingJob.builds;
+      if (_builds.isEmpty ||
+          _builds.first.number == buildingJob.firstBuild.number) {
+        builds = _builds;
+      } else {
+        final _earliestBuild = _builds.first;
+        final difference = _earliestBuild.number - build.buildNumber;
+
+        if (difference <= 0) {
+          builds = buildingJob.builds;
+        } else {
+          numberOfBuilds += difference;
+        }
+      }
+    }
+
+    return _processJenkinsBuilds(
+      builds,
+      buildingJob.name,
+      startFromBuildNumber: build.buildNumber,
+    );
+  }
+
+  @override
+  Future<List<BuildData>> fetchBuilds(String projectId) async {
+    final buildingJob = await _fetchBuilds(projectId);
+    return _processJenkinsBuilds(
+      buildingJob.builds,
+      buildingJob.name,
+    );
+  }
+
+  /// Fetches builds data for the project with given [projectId].
+  ///
+  /// The [limits] can be used to set a range-specifier for the request.
+  Future<JenkinsBuildingJob> _fetchBuilds(
+    String projectId, {
+    JenkinsQueryLimits limits = const JenkinsQueryLimits.empty(),
+  }) async {
+    final newBuildsFetchResult =
+        await jenkinsClient.fetchBuilds(projectId, limits: limits);
+    _throwIfInteractionUnsuccessful(newBuildsFetchResult);
+    return newBuildsFetchResult.result;
+  }
+
+  /// Checks the given [interactionResult] to be [InteractionResult.isSuccess].
+  ///
+  /// Throws [StateError] with the message of [interactionResult] if this
+  /// result is [InteractionResult.isError].
+  void _throwIfInteractionUnsuccessful(InteractionResult interactionResult) {
+    if (interactionResult.isError) {
+      throw StateError(interactionResult.message);
+    }
+  }
+
+  /// Processes the given [builds] to the list of [BuildData]s.
+  ///
+  /// The [jobName] is used to identify a building job for the builds.
+  /// The [startFromBuildNumber] is used to filter builds which number is less
+  /// than or equal to this value. This allows to avoid fetching old builds
+  /// since the range specifier in Jenkins API only provides ability to set
+  /// fetch limits but not to filter data to fetch.
```

Comment:
I'm not sure how we avoid fetching here as this piece called after fetching? Could you please explain this part a bit more?

Reply:
The Jenkins range-specifier works as a `sublist` for the resulting array-like property it is applied to. And not filters data. Therefore, the following situation is possible: 
- The last build in storage has the `buildNumber` equal to 10.
- There are builds with numbers 11, 12, 13, 14, 15 respectively. 
- A user deletes the builds with numbers 12 and 13. 
- The algorithm fetches the last build from CI - the result is 15. 
- Since the last build in CI is 15 and the last one in storage is 10, we should fetch 5 builds. 
- The algorithm fetches 5 builds - the result is 9, 10, 11, 14, 15. But we don't need to store 9 and 10, so we should filter.

So the range-specifier does not allow to handle such a situation and we have to add additional parameters.